library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationSource.java

```diff
@@ -89,17 +89,23 @@ void startSettingsApiResolutionForResult(Status status, Activity activity) throw
 
     @SuppressWarnings("ResourceType")
     void requestLocationUpdate() {
-        LocationServices.FusedLocationApi.requestLocationUpdates(googleApiClient, locationRequest, this);
+        if (isGoogleApiClientConnected()) {
+            LocationServices.FusedLocationApi.requestLocationUpdates(googleApiClient, locationRequest, this);
+        } else {
+            connectGoogleApiClient();
+            requestLocationUpdate();
+        }
     }
 
     void removeLocationUpdates() {
-        LocationServices.FusedLocationApi.removeLocationUpdates(googleApiClient, this);
+        if (isGoogleApiClientConnected())
```

Comment:
Only this line is necessary. And these methods should not interact with each other, so please try to do the check [before calling removeLocationUpdates](https://github.com/yayaa/LocationManager/blob/master/library/src/main/java/com/yayandroid/locationmanager/providers/locationprovider/GooglePlayServicesLocationProvider.java#L155) and cover it with test